packages/nimble-components/src/table-column/mixins/custom-sort-order.ts

```diff
@@ -0,0 +1,96 @@
+export function mixinCustomSortOrderColumnAPI<
+    TBase extends CustomSortOrderTableColumnConstructor<TColumnValidator>,
+    TColumnValidator extends ColumnValidator<['invalidCustomSortWithGrouping']>
+>(base: TBase) {
+    /**
+     * The Mixin that provides a concrete column with the API to allow sorting
+     * by a field other than the default for that column.
+     */
+    abstract class CustomSortOrderColumn extends base {
+        public sortByFieldName?: string;
+
+        /** @internal */
+        public notifier?: Notifier;
+
+        /** @internal */
+        public sortByFieldNameChanged(): void {
+            this.updateOperandDataRecordFieldName();
+
+            if (typeof this.sortByFieldName === 'string' && !this.notifier) {
+                this.notifier = Observable.getNotifier(this.columnInternals);
+                this.notifier.subscribe(this);
+                this.updateCustomColumnSortingValidity();
+            } else {
+                this.notifier?.unsubscribe(this);
+                this.notifier = undefined;
+                this.updateCustomColumnSortingValidity();
+            }
+        }
+
+        /** @internal */
+        public updateOperandDataRecordFieldName(): void {
+            if (typeof this.sortByFieldName === 'string') {
+                this.columnInternals.operandDataRecordFieldName = this.sortByFieldName;
+                this.columnInternals.sortOperation = TableColumnSortOperation.basic;
+            } else {
+                this.columnInternals.operandDataRecordFieldName = this.getDefaultSortFieldName();
+                this.columnInternals.sortOperation = this.getDefaultSortOperation();
+            }
+        }
+
+        /** @internal */
+        public handleChange(_source: unknown, args: unknown): void {
+            if (args === 'groupingDisabled' || args === 'groupIndex') {
+                this.updateCustomColumnSortingValidity();
+            } else if (args === 'dataRecordFieldNames') {
+                this.updateOperandDataRecordFieldName();
+            }
+        }
+
+        /** @internal */
+        public updateCustomColumnSortingValidity(): void {
+            const hasCustomColumnSorting = typeof this.sortByFieldName === 'string';
+            const isGrouped = !this.columnInternals.groupingDisabled
+                && typeof this.columnInternals.groupIndex === 'number';
+            const isValid = !isGrouped || !hasCustomColumnSorting;
+            if (isValid) {
+                this.columnInternals.validator.untrack(
+                    'invalidCustomSortWithGrouping'
+                );
+            } else {
+                this.columnInternals.validator.track(
+                    'invalidCustomSortWithGrouping'
+                );
+            }
+        }
+
+        /** @internal */
+        public abstract getDefaultSortFieldName(): string | undefined;
+
+        /** @internal */
+        public abstract getDefaultSortOperation(): TableColumnSortOperation;
```

Comment:
I think I find this pattern a bit confusing / challenging to follow. The contract of who owns what state and what to do when changes happen is a little unclear.

Didn't walkthrough it completely but my hunch is that it'll be a bit cleaner if the concrete / leaf class owns the configuration of all the columnInternals properties (operandDataRecordFieldName, sortOperation, fieldName) and is leveraging the mixin to ask for advice instead of the mixin configuring some of the state and the concrete class other bits of state.

Something like:
```ts
abstract class CustomSortOrderColumn extends base {
    public sortByFieldName?: string;

    public getResolvedOperandDataRecordFieldName (initialOperandFieldName) {}
    public getResolvedSortOperation (initialSortOperation) {}

    abstract handleCustomSortOrderConfigurationChanged{} /* definitely not committed to this name*/
    /* still handle validity stuff and attr handling stuff*/
}


class Text {
    protected override getColumnInternalsOptions():
        return {
            ...
            sortOperation: this.getResolvedSortOperation(TableColumnSortOperation.localeAwareCaseSensitive),
        };
    }
      
    public override handleCustomSortOrderConfigurationChanged (){
        this.columnInternals.dataRecordFieldNames = [this.fieldName] as const;
        this.columnInternals.operandDataRecordFieldName = this.getResolvedOperandDataRecordFieldName(this.filedName)
        this.columnInternals.sortOperation = this.getResolvedSortOperation(localAwareCaseInsenitive)
    }

    override fieldNameChanged(): void {
        this.columnInternals.dataRecordFieldNames = [this.fieldName] as const;
        this.columnInternals.operandDataRecordFieldName = this..getResolvedOperandDataRecordFieldName(this.filedName)
        this.columnInternals.sortOperation = this.getResolvedSortOperation(localAwareCaseInsenitive)
    }
   
}
```